best-practices/puppet_prs.md

```diff
@@ -33,3 +33,36 @@ git push origin sunetidFeature
 ```
 
  Next, create a PR in GitHub for your changes. Once the PR has been submitted, the ops team will be automatically notified to review it. Once the PR has been reviewed and merged into the production branch by the ops team, the puppetmaster will apply the configuration on its next run, usually within an hour, or a friendly operations team member will apply the change to the server by running the puppet agent manually.
+
+### Testing your puppet changes
+
+If you need to have the ability on the development machine to do a test run of the proposed puppet changes, you need to have access to the machine as the root user. Add `ksu: true` to the `app_user` configuration as shown:
+
+```
+role::params::app_user:
+  requests:
+    k5users:
+      - mysunetid
+      - yoursunetid
+    ksu:
+      true
+```
+
+Be aware that this will give root access to all of the `k5users`, so be certain that you actually need this level of access and that all of the k5users can have this level of access. Once the PR is submitted with this change the ops team will review and merge it as described above, and after the puppetmaster applies the configuration on the next run, the k5users will have root access to do test puppet runs.
+
+The next time you create a new puppet branch and commit/push to the puppet Github repository, you will be able to do a test run of the proposed puppet configuration to see what changes would actually be applied. To do this use the `ksu` command:
+
+```
+$ ksu
+Authenticated mysunetid@stanford.edu
+Account root: authorization for mysunetid@stanford.edu successful
+Changing uid to root (0)
+```
+
+Now you have the ability to do a test puppet run. This will run the puppet agent against the branch that you pushed to puppet, but will not actually apply any of the changes. You will see a diff of changes that will be made when the PR is finally merged and puppetmaster picks up and applies the change on your server:
+
+```
+$ puppet agent --test --noop environment=sunetidFeature
+```
+
+WARNING: Please make sure that you use the `--noop` flag or else the puppetmaster will <EM>always</EM> try to use the puppet configuration from the branch you created instead of production!
```

Comment:
Without the `--noop` flag, puppet will try to use the configuration from whatever environment is specified. If you run
```
puppet agent --test --environment=sunetidFeature
```
then puppet will apply the changes from the `sunetidFeature` branch.

If you run

```
puppet agent --test
```
then puppet will try to run in the environment specified in the configuration file at `/etc/puppetlabs/puppet/puppet.conf`

And if you ran the former command, puppet would update that setting in the config file with the environment you specified on the command line.

Reply:
I'm OK with this being in the same document. Changes requested are: making the `environment` parameter a flag, and clarifying which branch a run without `--noop` puppet will try to apply.